### PR TITLE
Add dns_data module

### PR DESCRIPTION
**Pull request by nrobinaubertin**
### Overview

The goal of this module is to regroup functions that handle the dns data but do not do any business operations on the charm.  
This is only moving lines of code, no logic has changed.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/dns_data.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        found_conflict = False
+        if entry in conflicting_entries:
+            continue
+        for e in entries:
+            if entry.conflicts(e) and entry != e:
+                found_conflict = True
+                conflicting_entries.add(entry)
+                conflicting_entries.add(e)
+
+        if not found_conflict:
```
Comment (arturo-seijas) [inline src/dns_data.py:164]: You can probably substract the set of conclicting entries from the list instead, which seems impler

File: src/dns_data.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    Returns:
+        A tuple containing the non-conflicting and conflicting entries
+    """
+    entries = [e for z in zones for e in z.entries]
+    nonconflicting_entries: set[models.DnsEntry] = set()
+    conflicting_entries: set[models.DnsEntry] = set()
+    while entries:
+        entry = entries.pop()
+        found_conflict = False
+        if entry in conflicting_entries:
```
Comment (arturo-seijas) [inline src/dns_data.py:156]: I think you can remember this and check if the entry is conflicting, unless there's a big performance penalty, oc

File: src/dns_data.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            continue
+        for e in entries:
+            if entry.conflicts(e) and entry != e:
+                found_conflict = True
+                conflicting_entries.add(entry)
```
Comment (arturo-seijas) [inline src/dns_data.py:161]:
```
conflicting_entries.add(entry)
conflicting_entries.add(e)
```
This is equivalent to `conflicting_entries.add(entry)` as `entry` will eventually have had the value for every entry in the list

**Review by javierdelapuente — APPROVED** [review]

File: src/dns_data.py (hunk 3 above)
Comment (nrobinaubertin) [reply src/dns_data.py:161]: After discussion with Arturo, I'm going to revisit this in a later PR

File: src/dns_data.py (hunk 2 above)
Comment (nrobinaubertin) [reply src/dns_data.py:156]: After discussion with Arturo, I'm going to revisit this in a later PR

File: src/dns_data.py (hunk 1 above)
Comment (nrobinaubertin) [reply src/dns_data.py:164]: After discussion with Arturo, I'm going to revisit this in a later PR

**Review by amandahla — APPROVED** [review]

**Pull request merged**